File: cprdspy/CPR_matplotlib/Spirals/spiral.py
import matplotlib.pyplot as plt
import numpy as np


def logSpiral(
    n=4,
    a=1,
    b=1,
    cyc=0.5,
    color="b",
    theta=0,
    rotation=0,
    direction="both",
    alpha=1,
    linewidth=1.0,
    use_degree=True,
):
    """绘制广义对数螺线。

    参数:
        n: 花瓣/对称数（影响基底 cos(pi/n)）
        a, b: x/y 方向缩放（可做椭圆化）
        cyc: 周期数（每个周期为 2π）
        color: 线颜色（传给 matplotlib）
        theta: 全局相位偏移（度或弧度取决于 use_degree）
        rotation: 额外角度偏移（度或弧度取决于 use_degree）
        direction: 'both'|'in'|'out' 控制绘制方向
        alpha: 透明度
        linewidth: 线宽
        use_degree (bool): theta 和 rotation 是否使用角度制，默认 True（度）。
    """
    theta_rad = np.deg2rad(theta) if use_degree else theta
    rot_rad = np.deg2rad(rotation) if use_degree else rotation

    if direction not in ("both", "in", "out"):
        raise ValueError("direction must be 'both', 'in' or 'out'")

    if direction == "both":
        t = np.linspace(-cyc * 2 * np.pi, cyc * 2 * np.pi, 1000)
    elif direction == "out":
        t = np.linspace(0, cyc * 2 * np.pi, 1000)
    else:  # in
        t = np.linspace(0, -cyc * 2 * np.pi, 1000)

    # r 随角度成几何增长/衰减：基于 (cos(pi/n)) ** (-n * t / pi)
    r = (np.cos(np.pi / n)) ** (-n * t / np.pi)
    ang = t + theta_rad + rot_rad
    x = a * r * np.cos(ang)
    y = b * r * np.sin(ang)

    plt.plot(x, y, color=color, alpha=alpha, linewidth=linewidth)
    plt.axis("equal")


def nSpiral(
    n=4,
    a=1,
    N=4,
    cyc=0.5,
    color="b",
    theta=0,
    rotation=0,
    direction="both",
    linewidth=1.0,
    use_degree=True,
):
    """n 边形螺旋。

    theta, rotation: 度或弧度取决于 use_degree。
    use_degree (bool): 默认 True（度）。
    """
    theta_rad = np.deg2rad(theta) if use_degree else theta
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(N):
        logSpiral(n, a, 1, cyc, color, theta_rad + i * 2 * np.pi / N, rot_rad,
                  direction, 1, linewidth, use_degree=False)
        logSpiral(n, a, 1, cyc, color, -theta_rad + i * 2 * np.pi / N, -rot_rad,
                  direction, 1, linewidth, use_degree=False)


def nSpirals(
    n=4,
    N=4,
    cyc=0.5,
    color="b",
    theta=0,
    rotation=0,
    direction="both",
    linewidth=1.0,
    use_degree=True,
):
    """多头 n 边形螺旋。

    theta, rotation: 度或弧度取决于 use_degree。
    use_degree (bool): 默认 True（度）。
    """
    theta_rad = np.deg2rad(theta) if use_degree else theta
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(N):
        logSpiral(n, 1, 1, cyc, color, theta_rad + i * 2 * np.pi / N, rot_rad,
                  direction, 1, linewidth, use_degree=False)
        logSpiral(n, -1, 1, cyc, color, -theta_rad + i * 2 * np.pi / N, -rot_rad,
                  direction, 1, linewidth, use_degree=False)


def calla_petal(
    n=4,
    a=1,
    cyc=1.25,
    theta=0,
    color="b",
    rotation=0,
    direction="both",
    linewidth=1.0,
    use_degree=True,
):
    """马蹄莲花瓣。

    theta, rotation: 度或弧度取决于 use_degree。
    use_degree (bool): 默认 True（度）。
    """
    theta_rad = np.deg2rad(theta) if use_degree else theta
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    logSpiral(n, a, 1, cyc, color, theta_rad, rot_rad, direction, 1, linewidth, use_degree=False)
    logSpiral(n, -a, 1, cyc, color, -theta_rad, -rot_rad, direction, 1, linewidth, use_degree=False)


def calla(
    n=4,
    a=1,
    cyc=1.25,
    N=12,
    theta=0,
    colors=["r", "g", "b"] * 4,
    rotation=0,
    direction="both",
    linewidth=1.0,
    use_degree=True,
):
    """完整马蹄莲。

    theta, rotation: 度或弧度取决于 use_degree。
    use_degree (bool): 默认 True（度）。
    """
    theta_rad = np.deg2rad(theta) if use_degree else theta
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(N):
        calla_petal(n, a, cyc, theta_rad + i * 2 * np.pi / N, colors[i],
                    rot_rad, direction, linewidth, use_degree=False)


# def logSpiral_in_out(n, a, b, cyc, color="b", theta=0):
#     t = np.linspace(-cyc * 2 * np.pi, cyc * 2 * np.pi, 1000)
#     x = a * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.cos(t + theta)
#     y = b * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.sin(t + theta)
#     plt.plot(x, y, color)
#     plt.axis("equal")


# def logSpiral_out(n, a, b, cyc, color="b", theta=0):
#     t = np.linspace(0, cyc * 2 * np.pi, 100)
#     x = a * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.cos(t + theta)
#     y = b * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.sin(t + theta)
#     plt.plot(x, y, color)


# def logSpiral_in(n, a, b, cyc, color="b", theta=0):
#     t = np.linspace(0, -cyc * 2 * np.pi, 100)
#     x = a * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.cos(t + theta)
#     y = b * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.sin(t + theta)
#     plt.plot(x, y, color)


# def n_spiral(n, cyc, color, theta=0):
#     for i in range(n):
#         logSpiral(n, 1, 1, cyc, color, theta + i * 2 * np.pi / n)
#         logSpiral(n, -1, 1, cyc, color, theta + i * 2 * np.pi / n)


# def n_spiral_rotate(n, cyc, color, alpha=0, theta=0):
#     for i in range(n):
#         logSpiral(n, 1, 1, cyc, color, alpha + theta + i * 2 * np.pi / n)
#         logSpiral(n, -1, 1, cyc, color, alpha - theta + i * 2 * np.pi / n)


# def n_spiral_rotate_out(n, cyc, color, theta=0):
#     for i in range(n):
#         logSpiral_out(n, 1, 1, cyc, color, theta + i * 2 * np.pi / n)
#         logSpiral_out(n, -1, 1, cyc, color, -theta + i * 2 * np.pi / n)


# def n_spiral_rotate_in(n, cyc, color, theta=0):
#     for i in range(n):
#         logSpiral_in(n, 1, 1, cyc, color, theta + i * 2 * np.pi / n)
#         logSpiral_in(n, -1, 1, cyc, color, -theta + i * 2 * np.pi / n)


# def calla_petal(n, cyc, theta, color):
#     logSpiral(n, 1, 1, cyc * 1.25, color, theta)
#     logSpiral(n, -1, 1, cyc * 1.25, color, -theta)


# def calla_by_petal(n, cyc, N, theta, colors):
#     for i in range(N):
#         calla_petal(n, cyc, theta + i * 2 * np.pi / N, colors[i])

File: cprdspy/CPR_js1/Spirals/spiral.js
/**
 * CPR 图形库 - JavaScript 版本
 * Spirals 模块 - 螺旋线相关函数
 */

/**
 * 绘制对数螺线
 * @param {Object} options - 配置选项
 * @param {number} options.n - 花瓣/对称数
 * @param {number} options.a - x 方向缩放
 * @param {number} options.b - y 方向缩放
 * @param {number} options.cyc - 周期数
 * @param {string} options.color - 颜色
 * @param {number} options.theta - 相位偏移
 * @param {number} options.rotation - 旋转角度
 * @param {string} options.direction - "both", "in", "out"
 * @param {number} options.alpha - 透明度
 * @param {number} options.lineWidth - 线宽
 * @returns {Object} 螺线路径数据
 */
function logSpiral({
    n = 4,
    a = 1,
    b = 1,
    cyc = 0.5,
    color = "b",
    theta = 0,
    rotation = 0,
    direction = "both",
    alpha = 1,
    lineWidth = 1,
} = {}) {
    if (!["both", "in", "out"].includes(direction)) {
        throw new Error("direction must be 'both', 'in' or 'out'");
    }

    let tStart, tEnd;
    if (direction === "both") {
        tStart = -cyc * 2 * Math.PI;
        tEnd = cyc * 2 * Math.PI;
    } else if (direction === "out") {
        tStart = 0;
        tEnd = cyc * 2 * Math.PI;
    } else {
        tStart = 0;
        tEnd = -cyc * 2 * Math.PI;
    }

    const points = 1000;
    const pathData = [];

    for (let i = 0; i <= points; i++) {
        const t = tStart + ((tEnd - tStart) * i) / points;
        const r = Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI);
        const ang = t + theta + rotation;
        const x = a * r * Math.cos(ang);
        const y = b * r * Math.sin(ang);

        if (i === 0) {
            pathData.push(`M ${x} ${y}`);
        } else {
            pathData.push(`L ${x} ${y}`);
        }
    }

    return {
        type: "logSpiral",
        d: pathData.join(" "),
        fill: "none",
        stroke: color,
        "stroke-width": lineWidth,
        "stroke-opacity": alpha,
    };
}

/**
 * 绘制 n 瓣螺旋
 */
function nSpiral({
    n = 4,
    N = 4,
    cyc = 0.5,
    color = "b",
    theta = 0,
    rotation = 0,
    lineWidth = 1,
} = {}) {
    const paths = [];

    for (let i = 0; i < N; i++) {
        const t1 = logSpiral({
            n,
            a: 1,
            b: 1,
            cyc,
            color,
            theta: theta + (i * 2 * Math.PI) / N,
            rotation,
            direction: "both",
            lineWidth,
        });

        const t2 = logSpiral({
            n,
            a: -1,
            b: 1,
            cyc,
            color,
            theta: -theta + (i * 2 * Math.PI) / N,
            rotation: -rotation,
            direction: "both",
            lineWidth,
        });

        paths.push(t1, t2);
    }

    return paths;
}

/**
 * 双向对数螺线
 */
function logSpiralInOut({
    n,
    a,
    b,
    cyc,
    color = "b",
    theta = 0,
} = {}) {
    const points = 1000;
    const tStart = -cyc * 2 * Math.PI;
    const tEnd = cyc * 2 * Math.PI;
    const pathData = [];

    for (let i = 0; i <= points; i++) {
        const t = tStart + ((tEnd - tStart) * i) / points;
        const r = Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI);
        const x = a * r * Math.cos(t + theta);
        const y = b * r * Math.sin(t + theta);

        if (i === 0) {
            pathData.push(`M ${x} ${y}`);
        } else {
            pathData.push(`L ${x} ${y}`);
        }
    }

    return {
        type: "logSpiral",
        d: pathData.join(" "),
        fill: "none",
        stroke: color,
    };
}

/**
 * 外向对数螺线
 */
function logSpiralOut({
    n,
    a,
    b,
    cyc,
    color = "b",
    theta = 0,
} = {}) {
    const points = 100;
    const pathData = [];

    for (let i = 0; i <= points; i++) {
        const t = (cyc * 2 * Math.PI * i) / points;
        const r = Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI);
        const x = a * r * Math.cos(t + theta);
        const y = b * r * Math.sin(t + theta);

        if (i === 0) {
            pathData.push(`M ${x} ${y}`);
        } else {
            pathData.push(`L ${x} ${y}`);
        }
    }

    return {
        type: "logSpiral",
        d: pathData.join(" "),
        fill: "none",
        stroke: color,
    };
}

/**
 * 内向对数螺线
 */
function logSpiralIn({
    n,
    a,
    b,
    cyc,
    color = "b",
    theta = 0,
} = {}) {
    const points = 100;
    const pathData = [];

    for (let i = 0; i <= points; i++) {
        const t = (-cyc * 2 * Math.PI * i) / points;
        const r = Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI);
        const x = a * r * Math.cos(t + theta);
        const y = b * r * Math.sin(t + theta);

        if (i === 0) {
            pathData.push(`M ${x} ${y}`);
        } else {
            pathData.push(`L ${x} ${y}`);
        }
    }

    return {
        type: "logSpiral",
        d: pathData.join(" "),
        fill: "none",
        stroke: color,
    };
}

/**
 * 绘制 n 个螺旋
 */
function nSpiralRotate({
    n,
    cyc,
    color,
    alpha = 0,
    theta = 0,
} = {}) {
    const paths = [];

    for (let i = 0; i < n; i++) {
        const t1 = logSpiral({
            n,
            a: 1,
            b: 1,
            cyc,
            color,
            theta: alpha + theta + (i * 2 * Math.PI) / n,
            lineWidth: 1,
        });

        const t2 = logSpiral({
            n,
            a: -1,
            b: 1,
            cyc,
            color,
            theta: alpha - theta + (i * 2 * Math.PI) / n,
            lineWidth: 1,
        });

        paths.push(t1, t2);
    }

    return paths;
}

/**
 * 旋转外向螺旋
 */
function nSpiralRotateOut({
    n,
    cyc,
    color,
    theta = 0,
} = {}) {
    const paths = [];

    for (let i = 0; i < n; i++) {
        const t1 = logSpiralOut({
            n,
            a: 1,
            b: 1,
            cyc,
            color,
            theta: theta + (i * 2 * Math.PI) / n,
        });

        const t2 = logSpiralOut({
            n,
            a: -1,
            b: 1,
            cyc,
            color,
            theta: -theta + (i * 2 * Math.PI) / n,
        });

        paths.push(t1, t2);
    }

    return paths;
}

/**
 * 旋转内向螺旋
 */
function nSpiralRotateIn({
    n,
    cyc,
    color,
    theta = 0,
} = {}) {
    const paths = [];

    for (let i = 0; i < n; i++) {
        const t1 = logSpiralIn({
            n,
            a: 1,
            b: 1,
            cyc,
            color,
            theta: theta + (i * 2 * Math.PI) / n,
        });

        const t2 = logSpiralIn({
            n,
            a: -1,
            b: 1,
            cyc,
            color,
            theta: -theta + (i * 2 * Math.PI) / n,
        });

        paths.push(t1, t2);
    }

    return paths;
}

/**
 * 马蹄莲花瓣
 */
function callaPetal({
    n,
    cyc,
    theta,
    color,
} = {}) {
    const t1 = logSpiral({
        n,
        a: 1,
        b: 1,
        cyc: cyc * 1.25,
        color,
        theta,
        lineWidth: 1,
    });

    const t2 = logSpiral({
        n,
        a: -1,
        b: 1,
        cyc: cyc * 1.25,
        color,
        theta: -theta,
        lineWidth: 1,
    });

    return [t1, t2];
}

/**
 * 多个马蹄莲花瓣
 */
function callaByPetal({
    n,
    cyc,
    N,
    theta,
    colors,
} = {}) {
    const paths = [];

    for (let i = 0; i < N; i++) {
        const petalColor = colors[i % colors.length];
        const petals = callaPetal({
            n,
            cyc,
            theta: theta + (i * 2 * Math.PI) / N,
            color: petalColor,
        });
        paths.push(...petals);
    }

    return paths;
}

// 导出模块
export {
    logSpiral,
    logSpiralOut,
    logSpiralIn,
    logSpiralInOut,
    nSpiral,
    nSpiralRotate,
    nSpiralRotateOut,
    nSpiralRotateIn,
    callaPetal,
    callaByPetal,
};
